### DISCUSSION TITLE
What's a normal CPU load?

### DISCUSSION THREAD

**Discussion by Albarge** [Louisvdw/dbus-serialbattery#1033] [General]
I think my Cerbo GX is struggling, I was wondering what other peoples CPU's are like whether Pi or Cerbo etc to compare to?

I think mine seems a bit high and have found the later versions of SerialBattery to overload the system and shut off the power, so before I try the latest stable v1.2.20240408 I wanted to idiot check that my Cerbo can handle it. I tried v1.2.20240227beta, and some Nightly's but had to stick with v1.0.20240102beta as the latest version that was stable for me.

I am using GuiMods, BatteryAggregator and I have 3 Waveshare industrial serial adaptors/batteries, 2 SmartShunts, 3 Ve.Can MPPTS, 1 VeDirect MPPT, 2 Bluetooth Dongles, 2 USB hubs, 4 RuuviTags, 1 digital input for generator signal, 1 resistive tank sender. But don't use MQTT.

Here's my htop earlier today, Edit* I wasn't looking properly and had thought the top process line in light blue was the overall CPU load, but realise it's the two cores 1+2 at the top. I regularly see them 70%-99.5% 

![htop 97%](https://github.com/Louisvdw/dbus-serialbattery/assets/98849340/70533bd7-f6ea-44f7-8604-1c4ed5be0b5c)

Is it normal? What are others like? Do I have any processes that look bad?

**Comment by coca7** [comment]
I used GuiMods and it really showed down my Cerbo.  I uninstalled it and now the system runs much better.

**Reply by Albarge** [reply]
Yeah, It does also use resources, not sure everything in the htop list is necessary or efficient! I hope I can keep GuiMods though, although I can see I might have to uninstall it when Venus GUI V2 is mature, but the new brief and overview pages are useless for information, I love just being able to glance at GuiMods and see so much critical info on one page.

**Reply by Albarge** [reply]
I just looked again today, and edited my first post as I was being dumb, and hadn't realised the two processor cores were maxing 70%-99.5% most of the time. Do I have suspect processes? 

Or is it just GuiMods and Dbus-SerialBattery with BatteryAggregator can't run together on the Cerbo GX?

![Screenshot 2024-04-11 141404](https://github.com/Louisvdw/dbus-serialbattery/assets/98849340/b9ca27e6-e69e-4a36-a30b-853f0f769489)

**Comment by mr-manuel** [comment]
I checked a bit the default config (of the latest version) and what options can be used to save a bit performance. Change this settings to this values and check, if it helps:

* `LINEAR_LIMITATION_ENABLE = False`
  OR
  `LINEAR_LIMITATION_ENABLE = True`
  `LINEAR_RECALCULATION_EVERY = 300`
  `LINEAR_RECALCULATION_ON_PERC_CHANGE = 20`
* `TIME_TO_GO_ENABLE = False`
* `PUBLISH_CONFIG_VALUES = False`

But I don't think that the gain is huge.

You could also sort by `TIME+` in htop to see which process used most CPU time.

**Reply by Albarge** [reply]
Thanks I realise it's probably not strictly a serialbattery issue, but I figured yourself and the folks looking here might have some useful insights.

I'll try the last two suggestions in a bit, I already found: `LINEAR_LIMITATION_ENABLE = True` made the system too unstable, so I use step mode.

My TIME+: I'm not sure what aspects are GuiMods, but Setuphelper/packagemanager and BatteryAggregator use up a bit of processing, whereas the serialbattery instances are lower down the list. Before hitting `CELL_VOLTAGES_WHILE_CHARGING` the CPU was between 50-70%:

![TIME+](https://github.com/Louisvdw/dbus-serialbattery/assets/98849340/9c8b5535-9870-41d3-9cf1-eccae4c88c4f)

I'm a bit surprised I'm often so close to the limits of CPU power, as I replaced a CCGX running GuiMods before my Pb to LFP upgrade assuming the Cerbo GX would be future proof for a while, and I'm not using OS Large or Node Red either.

My config running on v1.0.20240102beta:

```ini
MAX_BATTERY_CHARGE_CURRENT    = 170.0
MAX_BATTERY_DISCHARGE_CURRENT = 200.0

MIN_CELL_VOLTAGE   = 2.900

MAX_CELL_VOLTAGE   = 3.500

FLOAT_CELL_VOLTAGE = 3.375

BLOCK_ON_DISCONNECT = False

LINEAR_LIMITATION_ENABLE = False

CVCM_ENABLE = True

MAX_VOLTAGE_TIME_SEC = 3600

SOC_LEVEL_TO_RESET_VOLTAGE_LIMIT = 90

CCCM_CV_ENABLE = True

CELL_VOLTAGES_WHILE_CHARGING   = 3.505, 3.49, 3.48, 3.47, 3.46, 3.45, 3.30
MAX_CHARGE_CURRENT_CV_FRACTION =    0,  0.1,  0.2,  0.3,  0.5,  0.5,    1


DCCM_CV_ENABLE = False

CELL_VOLTAGES_WHILE_DISCHARGING   = 2.70, 2.80, 2.90, 3.10
MAX_DISCHARGE_CURRENT_CV_FRACTION =    0,  0.1,  0.5,    1


TEMPERATURE_LIMITS_WHILE_CHARGING = 0,   5,  10,  15, 20, 35,  40, 50, 55
MAX_CHARGE_CURRENT_T_FRACTION     = 0,  0.3, 0.6, 1,  1,  1,   1,  0.6, 0

TEMPERATURE_LIMITS_WHILE_DISCHARGING  = -20,   0,   5,  10, 15, 45, 55
MAX_DISCHARGE_CURRENT_T_FRACTION     =   0, 0.4, 0.6, 0.8,  1,  1,  0


CCCM_SOC_ENABLE = False

DCCM_SOC_ENABLE = False


TIME_TO_SOC_POINTS = 100, 0

TIME_TO_SOC_VALUE_TYPE = 2

TIME_TO_SOC_RECALCULATE_EVERY = 120


BMS_TYPE = Jkbms


TEMP_BATTERY = 0


TEMP_1_NAME = Bottom


TEMP_2_NAME = Top


SOC_LOW_WARNING = 0
SOC_LOW_ALARM   = 5
```

**Reply by Albarge** [reply]
With GuiMods Uninstalled it's still hitting 100% CPU, so is this normal for others too? Or I have an issue, what should I be hoping for max CPU% ? Maybe I need to try factory reset of Cerbo with a fresh Venus v3.30?

After uninstall of GuiMods, so serialbattery and batteryaggregator the only other mods:
![GuiMods Uninstalled](https://github.com/Louisvdw/dbus-serialbattery/assets/98849340/7c6a4f3c-4985-4b86-9392-320a25b28d44)

**Reply by Honusnap** [reply]
Your vrmlogger seems to tazke far too high CPU
Here is my htop 👍 
![2024-04-15_105353](https://github.com/Louisvdw/dbus-serialbattery/assets/507094/1a738c34-39dc-464e-886c-5cfc6d045f28)

**Reply by Honusnap** [reply]
Got a question though .. why the watchdog take so much CPU .. ? Is that an artefact in htop or in the way those 2 process are connected ? Or does dserial really take 14+14 = 30% CPU ? Is the watchdog really taking 14% ,or is is something like it's a fork or whatever and system double count the CPU charge ?
Dserial do not seems to crash on my system, so .. if it take 14%, better make it an option .. :D

**Reply by coca7** [reply]
Here is my system.  I am also running serial battery and battery aggregator.  
<img width="1159" alt="Screenshot 2024-04-15 at 7 30 53 AM" src="https://github.com/Louisvdw/dbus-serialbattery/assets/2902641/5a237325-36be-4032-8385-c13bb851ad8c">

**Comment by Albarge** [comment]
Ok, I should probably make a thread on Victron community, but to update here, I decided the only real option was do a fresh 'install' so I've had a painful evening spending hours doing a reset to factory defaults: [https://www.victronenergy.com/media/pg/Cerbo_GX/en/reset-to-factory-defaults-and-venus-os-reinstall.html](url) Then reinstall and set up everything and fix all duplicated sensors and broken VRM widgets etc.. very boring.

Weird things happened during the process, I booted from v.3.30 with the factory file on USB and it didn't boot up properly; blank screen, and 3 weird beeps on loop (not the normal beep), I pulled out the USB and it sort of loaded, but only a few sections of the screen initially, then a reboot showed a mouse curser on the screen? 

I figured loads of other people are also having issues with VenusOS v3.30 so I booted the stored version v3.22 in the Cerbo, that seemed to work. Then slowly putting everything back on and renaming defaults, Ruuvitags, battery Aggregator, SerialBattery v1.2.20240401 Louis stable and finally GuiMods. I watched the `top` output slowly go up, whilst adding more functions back to Venus, but weirdly `htop` doesn't work now with v3.22? "command not found"

So far my top is looking better and can run the latest stable SerialBattery which I couldn't before, It's late and no sun yet, but hopefully the roughly 65% average CPU (35% idle) doesn't go up much more with the sun at absorption, at least it's looking closer to coca7 now:

![After complete resetup](https://github.com/Louisvdw/dbus-serialbattery/assets/98849340/2996a9a6-9b9c-4063-b52e-db98fb56cffb)

I still think there's an issue with corrupted files somewhere, whether that's something in Venus or from remnants of past corrupted  Mod files I don't know, but it wasn't a nice clean install, although so far it's running better.

**Reply by mr-manuel** [reply]
htop is not part of Venus OS. You have to install it: https://github.com/mr-manuel/venus-os_helpful-scripts/tree/master/htop/armv7

**Reply by Albarge** [reply]
Oh thanks, I must have installed it from the troubleshoot page before, but don't remember that!

I'm now getting average 60-70% CPU, which was better than before, but it does still spike to 98-100% every so often, but not as constantly close to it. I'm still yet to turn on linear limitation to see if the Cerbo can handle it, but maybe I'm just at the Cerbo max power.. : (

*Edit I'm running with Linear Limitation for a few days ok now, but still CPU is too high and laggy but stable.

![60to70percentaverage](https://github.com/Louisvdw/dbus-serialbattery/assets/98849340/a2265ff0-8eda-4dc5-9192-7f07972f5476)

**Reply by Honusnap** [reply]
Wait .. how dbus-Serial take 3.5% in this last screenshot, mine is at 16% .. !
Wow ... those discrepency are strange .. I use MQQT while seems you do not, i got a Bluetooth BMS connection, maybe you use CAN ? Would it explain this 13% CPU load difference ?

**Reply by Albarge** [reply]
I use USB to RS485 for each of my three batteries. But yeah they're normally around 1.7% -  4.8% CPU each.

But GUI is 30%-60% CPU! My system is mostly over the recommended max load average of 2 for a 2 core CPU though, I think I'm just asking too much from the Cerbo, so I'm going to get a rpi4 at some point soon.

**Reply by Honusnap** [reply]
Yea your display seems high... maybe you have selecetd too many options to display, i'm at 15% CPU for the display.

So ... the RS485 take ... let's say 2% per battery while i got only one by bluetooth and it take 15% CPU ... 
Clearly then the CPU is wasted on Bluetooth.
Bluetooth take 7x more CPU then RS485 ....  HELLLPPPPPP ... ;D

**Reply by Albarge** [reply]
Can you try TTL / RS485 to USB? or CAN?

**Reply by Honusnap** [reply]
Not easily, i think i got all needed (RS485 adapter for Jkbms and the compatible BMS) but it's complicated to get the cable through the wall .. :D 
But maybe i'll do that after tommorow.. !

Wait .. i did not ask what BMS are you using...? Cause maybe that's a reason for this Load difference.

**Reply by Albarge** [reply]
If you use the TTL to RS485 adaptor from JK, you still need an USB to RS485 (I use Waveshare Isolated adaptors after testing cheap ones that seemed to cause more errors) The Waveshare I used will also directly connect via TTL, so I didn't need the JK adaptor.

Make sure to use twisted pairs in your RS485 cable

I use this: https://amzn.eu/d/6BkqMBC

I have JK B2A8S20P V.11

**Reply by Honusnap** [reply]
I got a RS485 to USB, i just need to find time to do it... :D

**Reply by Albarge** [reply]
@Honusnap hopefully you do have everything, but it's not clear.

JK's don't output a standard RS485 (even though it's printed on them) they use TTL, that's why you need either TTL to USB or the JK RS485 adaptor (TTL to RS485) AND then another RS485 to USB adaptor.

If you don't have one of these (one is for 48v and the other for <24V) your RS485 to USB won't work: https://www.aliexpress.com/item/1005003451100409.html

![JKBMS RS485 CAN module and LCD display Adapter - AliExpress](https://github.com/Louisvdw/dbus-serialbattery/assets/98849340/2c6cb945-3fcc-4d30-ac85-b67b847fb7cb)

**Reply by Honusnap** [reply]
I got everything that needed, Jk adapter and RS485 to USB.
